### PR TITLE
feat: support custom webhooks in auth0 source transformer

### PR DESCRIPTION
**Pull request by mihir-4116**
## Description of the change

- previously we were using [Auth0 Authentication API Webhooks](https://auth0.com/docs/customize/extensions/auth0-authentication-api-webhooks) which is deprecated as of now. so the new way to create webhook is through [custom webhooks](https://auth0.com/docs/customize/log-streams/custom-log-streams)
- since both the types are webhooks there will not much change in webapp and config-be, but the structure of response is different for both types of webhooks. 
- in this pr auth0 code is refactored according to the custom webhooks(new way to configure webhook from auth0 dashboard) response structure

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ✅] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ✅] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Comment by saikumarrs** [comment]
The PR title says "fix" and "refactor". What is it exactly?

**Comment by saikumarrs** [comment]
If it's indeed a fix, can you add more details of the fix to the description to make it clear?

File: v0/sources/auth0/util.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+ * This function returns the organization_id from apiUrl path
+ * example
+    path : "/api/v2/organizations/org_DqzHLxgABzej3mwL/members"
+    groupId : org_DqzHLxgABzej3mwL
+ * @param {*} event
+ */
+const getGroupId = event => {
+  const { path } = event.details.request;
+  const pathArray = path.split("/");
+  return pathArray[pathArray.length - 2];
```
Comment (saikumarrs) [inline v0/sources/auth0/util.js:11]: What if any of the fields in the path `event.details.request.path` is undefined?

File: v0/sources/auth0/transform.js [hunk 2]
```diff
@@ -12,72 +13,68 @@ const eventNameMap = JSON.parse(
...
+  }
+  return prepareTrackPayload(event);
 }
 
 function process(events) {
   const responses = [];
   let eventList = events;
   if (!Array.isArray(events)) {
-    eventList = [events];
+    eventList = events.logs ? events.logs : [events];
```
Comment (saikumarrs) [inline v0/sources/auth0/transform.js:68]:
```suggestion
    eventList = events.logs || [events];
```

File: v0/sources/auth0/transform.js [hunk 3]
```diff
@@ -12,72 +13,68 @@ const eventNameMap = JSON.parse(
...
-        responses.push(removeUndefinedAndNullValues(resp));
-      }
-    } catch (error) {
-      // TODO: figure out a way to handle partial failures within batch
-      // responses.push({
-      //   statusCode: 400,
-      //   error: error.message || "Unknwon error"
-      // });
+    const resp = processEvent(event.data);
+    if (resp && resp.userId && resp.userId !== "") {
```
Comment (saikumarrs) [inline v0/sources/auth0/transform.js:72]: You can simplify and also a redundant check.
```suggestion
    if (resp?.userId) {
```

Comment (saikumarrs) [inline v0/sources/auth0/transform.js:72]: Why don't we send the entire `event` to `processEvent` and let it do this update as well?

File: v0/sources/auth0/transform.js [hunk 4]
```diff
@@ -12,72 +13,68 @@ const eventNameMap = JSON.parse(
...
+  // ss -> successful signup
+  if (eventType === "ss") {
+    return prepareIdentifyPayload(event);
+  }
+  // sapi -> Success API Operation
+  if (eventType === "sapi") {
+    if (
+      event.description &&
+      event.description === "Add members to an organization"
+    ) {
```
Comment (saikumarrs) [inline v0/sources/auth0/transform.js:56]: Redundant.
```suggestion
    if (event.description === "Add members to an organization") {
```

**Comment by saikumarrs** [comment]
> If it's indeed a fix, can you add more details of the fix to the description to make it clear?

From the changeset, it appears to be a feature addition supporting different Auth0 event types.

File: v0/sources/auth0/transform.js (hunk 4 above)
Comment (mihir-4116) [reply v0/sources/auth0/transform.js]: updated

File: v0/sources/auth0/transform.js (hunk 3 above)
Comment (mihir-4116) [reply v0/sources/auth0/transform.js]: updated

Comment (mihir-4116) [reply v0/sources/auth0/transform.js]: updated

File: v0/sources/auth0/transform.js (hunk 2 above)
Comment (mihir-4116) [reply v0/sources/auth0/transform.js]: updated

File: v0/sources/auth0/util.js (hunk 1 above)
Comment (mihir-4116) [reply v0/sources/auth0/util.js:12]: discussed this with @shrouti1507 and the final outcome is we will drop event if path is not there, i made changes according to that

**Review by shrouti1507 — DISMISSED** [review]

File: v0/sources/auth0/transform.js (hunk 4 above)
Comment (shrouti1507) [reply v0/sources/auth0/transform.js]: let's add optional chaining here, `event?.description`

File: v0/sources/auth0/transform.js [hunk 5]
```diff
@@ -38,43 +38,44 @@ const prepareGroupPayload = event => {
...
-    return prepareIdentifyPayload(event);
-  }
-  // sapi -> Success API Operation
-  if (eventType === "sapi") {
-    if (
-      event.description &&
-      event.description === "Add members to an organization"
-    ) {
-      return prepareGroupPayload(event);
+function processEvent(eventList) {
```
Comment (saikumarrs) [inline v0/sources/auth0/transform.js:41]: Perhaps this function name should be changed to `processEvents` now.

Comment (saikumarrs) [reply v0/sources/auth0/transform.js:41]: I'd still recommend restricting it to just handle a single event though.

Comment (mihir-4116) [reply v0/sources/auth0/transform.js]: function name updated to `processEvents`

**Review by saikumarrs — APPROVED** [review]

**Review by shrouti1507 — DISMISSED** [review]

**Review by Gauravudia — APPROVED** [review]

**Review by saikumarrs — DISMISSED** [review]

**Review by shrouti1507 — APPROVED** [review]

**Review by sauravlal15 — APPROVED** [review]

**Pull request merged**